src/middlewared/middlewared/plugins/filesystem.py

```diff
@@ -9,19 +9,31 @@
 import pathlib
 
 from middlewared.event import EventSource
-from middlewared.schema import accepts, Bool, Dict, Float, Int, List, Ref, returns, Path, Str
-from middlewared.service import private, CallError, filterable_returns, Service, job
-from middlewared.utils import filter_list
 from middlewared.plugins.pwenc import PWENC_FILE_SECRET
 from middlewared.plugins.cluster_linux.utils import CTDBConfig, FuseConfig
 from middlewared.plugins.filesystem_ import stat_x
+from middlewared.schema import accepts, Bool, Dict, Float, Int, List, Ref, returns, Path, Str
+from middlewared.service import private, CallError, filterable_returns, Service, job
+from middlewared.utils import filter_list, run
 
 
 class FilesystemService(Service):
 
     class Config:
         cli_namespace = 'storage.filesystem'
 
+    @private
+    async def add_immutable_flag_to_path(self, path):
+        cp = await run(['chattr', '+i', '-RV', path], check=False)
+        if cp.returncode:
```

Comment:
Why chattr and not `os.chflags()`? 

Why are we doing this recursively?

Reply:
We are doing this recursively to make sure that if there are any files in the directory, they cannot be modified as well.

Reply:
If it's a dataset mountpoint, shouldn't the dir be empty? I'm somewhat leery of recursive ops because we've been burned there before.

Reply:
Ideally yes but that wouldn't be true for any writes which may already have been done. The idea is to make sure that the complete path is immutable.

Reply:
> Ideally yes but that wouldn't be true for any writes which may already have been done. 
That sounds like a separate error condition. Perhaps on dataset lock / unmount, we check for whether stuff is inside that dir and generate an alert so that the user can take care of it.

Reply:
If it's a non-root user, then writes to sub-files/dirs will fail if top-level dir is immutable. If it's a root account, they will work.